File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* nodesBetweenCriticalPoints(struct ListNode* head, int* returnSize) {
    *returnSize = 2;
    int* answer = (int*)malloc((*returnSize) * sizeof(int));
    int criticalPoints = 0;
    answer[0] = INT_MAX;
    answer[1] = INT_MIN;
    int first_cp = -1;
    int prev_cp = 0;
    int pos = 2;

    struct ListNode* prv = head;
    struct ListNode* curr = head->next;
    struct ListNode* nxt = head->next->next;
    while (nxt) {
        if ((prv->val < curr->val && curr->val > nxt->val) || (prv->val > curr->val && curr->val < nxt->val)) {
            criticalPoints++;
            if (first_cp < 0) {
                first_cp = pos;
            }
            else {
                if (pos-first_cp > answer[1]) answer[1] = pos - first_cp;
                if (pos-prev_cp < answer[0]) answer[0] = pos - prev_cp;
            }
            prev_cp = pos;
        }
        prv = curr;
        curr = nxt;
        nxt = nxt->next;
        pos++;
    }
    if (criticalPoints < 2) {
        answer[0] = -1;
        answer[1] = -1;
    }

    return answer;
}

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.py
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def nodesBetweenCriticalPoints(self, head):
        prev = head
        curr = head.next
        nextt = head.next.next
        critical_points = []
        index = 1

        while nextt:
            if (prev.val < curr.val and curr.val > nextt.val) or (prev.val > curr.val and curr.val < nextt.val):
                critical_points.append(index)
            prev = curr
            curr = nextt
            nextt = nextt.next
            index += 1

        if len(critical_points) < 2:
            return [-1,-1]
        print(critical_points)
        min_distance = float('inf')
        max_distance = critical_points[-1] - critical_points[0]
        for i in range(len(critical_points)-1):
            for j in range(i+1, len(critical_points)):
                distance = critical_points[j]-critical_points[i]
                if distance < min_distance:
                    min_distance = distance 
        return [min_distance, max_distance]